# AAI612: Deep Learning & its Applications

*Notebook 5.4: Predicting Housing Prices*

<a href="https://colab.research.google.com/github/harmanani/AAI612/blob/main/Week5/Notebook5.4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
"""
The MIT License (MIT)
Copyright (c) 2021 NVIDIA
Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""


## Part I: Create the model

This example uses a neural network to solve a regression problem, using the Boston housing dataset.  The Boston Housing dataset is included in Keras, so it is simple to access using keras.datasets.boston_housing:

In [1]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import numpy as np

# Read and standardize the data.
boston_housing = keras.datasets.boston_housing
(raw_x_train, y_train), (raw_x_test,
    y_test) = boston_housing.load_data()

57026/57026 ━━━━━━━━━━━━━━━━━━━━ 0s 4us/step


We standardize both the training and test data by using the mean and standard deviation from the training data. The parameter axis=0 ensures that we compute the mean and standard deviation for each input variable separately. The resulting mean (and standard deviation) is a vector of means instead of a single value. That is, the standardized value of the nitric oxides concentration is not affected by the values of the per capita crime rate or any of the other variables.

In [2]:
x_mean = np.mean(raw_x_train, axis=0)
x_stddev = np.std(raw_x_train, axis=0)
x_train =(raw_x_train - x_mean) / x_stddev
x_test =(raw_x_test - x_mean) / x_stddev

Create the model by first instantiating the model object without any layers, and then add them one by one using the member method `add()`.
Next, complete the missing code in the below hidden layers using 64 ReLU neurons per layer.  Careful regarding the first layer as it needs to match the match the dataset. The output layer consists of a single neuron with a linear activation function. 

In [ ]:
# Create and train model.
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=[raw_x_train.shape[1]]))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation='linear'))

C:\Users\acer\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Use MSE as the loss function and use the `Adam` optimizer and compile method that we are interested in seeing the metric mean absolute error, and print out a summary of the model with `model.summary()` and then start training.  Experiment with different `batch sizes` and `epochs`.  Record the results for these experiments.

In [6]:
BATCH_SIZE = 16
EPOCHS = 500

model.compile(loss='mse', optimizer='Adam', metrics =['mean_absolute_error'])
model.summary()
history = model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=EPOCHS, batch_size=BATCH_SIZE, verbose=2, shuffle=True)

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 64)             │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │         4,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,121 (20.00 KB)

 Trainable params: 5,121 (20.00 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/500


C:\Users\acer\AppData\Roaming\Python\Python312\site-packages\keras\src\ops\nn.py:907: UserWarning: You are using a softmax over axis -1 of a tensor of shape (None, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(


26/26 - 2s - 64ms/step - loss: 542.3705 - mean_absolute_error: 21.3950 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 2/500
26/26 - 0s - 7ms/step - loss: 542.3704 - mean_absolute_error: 21.3950 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 3/500
26/26 - 0s - 7ms/step - loss: 542.3704 - mean_absolute_error: 21.3950 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 4/500
26/26 - 0s - 12ms/step - loss: 542.3704 - mean_absolute_error: 21.3951 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 5/500
26/26 - 0s - 7ms/step - loss: 542.3704 - mean_absolute_error: 21.3951 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 6/500
26/26 - 0s - 13ms/step - loss: 542.3704 - mean_absolute_error: 21.3950 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 7/500
26/26 - 0s - 7ms/step - loss: 542.3704 - mean_absolute_error: 21.3950 - val_loss: 570.7009 - val_mean_absolute_error: 22.0784
Epoch 8/500
26/26 - 0s - 6ms/step - loss: 5

After the training is done, we use our model to predict the price for the entire test set.

In [7]:
predictions = model.predict(x_test)

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step


C:\Users\acer\AppData\Roaming\Python\Python312\site-packages\keras\src\ops\nn.py:907: UserWarning: You are using a softmax over axis -1 of a tensor of shape (32, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(


Knowing the answers, let us check how far off you are:

In [8]:
expected =[7.939793, 18.455063, 20.115505, 32.14037]
for i in range(0, 4):
    assert ((predictions[i] - expected[i]) < 0.0001), "predicted value is too large"

## Part II: Tuning

Improve the above results by modifying the network to include more layers with more neurons.  Also, apply L1 and L2 regularization using different weight decay parameters: Start with $\lambda=0.1$ and try $\lambda=0.2$ and $\lambda=0.3$

In [15]:
from tensorflow.keras.layers import Dropout
from tensorflow.keras.regularizers import l1_l2
# Create and train model.

LAMBDA = 0.2

model = Sequential([
    Dense(128, activation='relu', kernel_regularizer=l1_l2(LAMBDA, LAMBDA), input_shape=[raw_x_train.shape[1]]),
    Dropout(0.3),
    Dense(128, activation='relu', kernel_regularizer=l1_l2(LAMBDA, LAMBDA)),
    Dropout(0.3),
    Dense(64, activation='relu', kernel_regularizer=l1_l2(LAMBDA, LAMBDA)),
    Dropout(0.3),
    Dense(64, activation='relu', kernel_regularizer=l1_l2(LAMBDA, LAMBDA)),
    Dense(1, activation='linear')
])

model.compile(optimizer='adam', loss='mse', metrics=['mae'])

model.summary()
history = model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=EPOCHS, batch_size=BATCH_SIZE, verbose=2, shuffle=True)

predictions = model.predict(x_test)

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_16 (Dense)                │ (None, 128)            │         1,792 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_8 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_17 (Dense)                │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_9 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_18 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_10 (Dropout)            │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_19 (Dense)                │ (None, 64)             │         4,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_20 (Dense)                │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 30,785 (120.25 KB)

 Trainable params: 30,785 (120.25 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/500
26/26 - 3s - 105ms/step - loss: 1075.2307 - mae: 21.3827 - val_loss: 930.9005 - val_mae: 19.1067
Epoch 2/500
26/26 - 0s - 7ms/step - loss: 663.4443 - mae: 11.1894 - val_loss: 499.3057 - val_mae: 6.8826
Epoch 3/500
26/26 - 0s - 6ms/step - loss: 461.2032 - mae: 6.3376 - val_loss: 392.5522 - val_mae: 4.5877
Epoch 4/500
26/26 - 0s - 7ms/step - loss: 381.9635 - mae: 5.5798 - val_loss: 331.5688 - val_mae: 4.1369
Epoch 5/500
26/26 - 0s - 14ms/step - loss: 321.2321 - mae: 4.9724 - val_loss: 278.1147 - val_mae: 3.5817
Epoch 6/500
26/26 - 0s - 8ms/step - loss: 272.3720 - mae: 4.6274 - val_loss: 240.3749 - val_mae: 3.6026
Epoch 7/500
26/26 - 0s - 6ms/step - loss: 241.0243 - mae: 4.6089 - val_loss: 214.9627 - val_mae: 3.8533
Epoch 8/500
26/26 - 0s - 8ms/step - loss: 211.8093 - mae: 4.3601 - val_loss: 195.0934 - val_mae: 4.1589
Epoch 9/500
26/26 - 0s - 7ms/step - loss: 192.8360 - mae: 4.3042 - val_loss: 176.1625 - val_mae: 3.8285
Epoch 10/500
26/26 - 0s - 7ms/step - loss: 173.6117 - mae

Retry the above using dropout regularization.  What do you notice?

Repeat the above by trying multiple parameter combinations:
- Using a combination of 64 to 128 neurons 
- Using Dropout of 0.2 and 0.3
- Using L2 = 0.1 and 0.2

For each case, print first 4 predictions and record the answers!

In [ ]:
for i in range(0, 4):
    print('Prediction: ', predictions[i, 0],
          ', true value: ', y_test[i])

Prediction:  15.898713 , true value:  7.2
Prediction:  18.259224 , true value:  18.8
Prediction:  20.246407 , true value:  19.0
Prediction:  33.0872 , true value:  27.0


lamda: 0.1
dropout:0.2
Prediction:  13.94176 , true value:  7.2
Prediction:  19.111694 , true value:  18.8
Prediction:  20.371693 , true value:  19.0
Prediction:  34.412743 , true value:  27.0

lamda:0.2
dropout: 0.3
Prediction:  15.898713 , true value:  7.2
Prediction:  18.259224 , true value:  18.8
Prediction:  20.246407 , true value:  19.0
Prediction:  33.0872 , true value:  27.0

after increasing the layers ther is a defenit increase in the performance 
as for lamda and the dropout experiment (lamda 0.1 and dropout 0.3) is better than the first trial 